enterprise/kernel/src/main/java/org/neo4j/kernel/impl/pagecache/WarmupAvailabilityListener.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2002-2018 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.pagecache;
+
+import java.util.concurrent.TimeUnit;
+
+import org.neo4j.graphdb.factory.GraphDatabaseSettings;
+import org.neo4j.kernel.AvailabilityGuard.AvailabilityListener;
+import org.neo4j.kernel.configuration.Config;
+import org.neo4j.logging.Log;
+import org.neo4j.scheduler.JobScheduler;
+
+import static org.neo4j.scheduler.JobScheduler.Groups.pageCacheIOHelper;
+
+class WarmupAvailabilityListener implements AvailabilityListener
+{
+    private final JobScheduler scheduler;
+    private final PageCacheWarmer pageCacheWarmer;
+    private final Config config;
+    private final Log log;
+    private final PageCacheWarmerMonitor monitor;
+
+    // We use the monitor lock to guard the job handle. However, it could happen that a job has already started, ends
+    // up waiting for the lock while it's being held by another thread calling `unavailable()`. In that case, we need
+    // to make sure that the signal to stop is not lost. Cancelling a job handle only works on jobs that haven't
+    // started yet, since we don't propagate an interrupt. This is why we check the `available` field in the
+    // `scheduleProfile` method.
+    private volatile boolean available;
+    private JobScheduler.JobHandle jobHandle; // Guarded by `this`.
+
+    WarmupAvailabilityListener( JobScheduler scheduler, PageCacheWarmer pageCacheWarmer,
+                                Config config, Log log, PageCacheWarmerMonitor monitor )
+    {
+        this.scheduler = scheduler;
+        this.pageCacheWarmer = pageCacheWarmer;
+        this.config = config;
+        this.log = log;
+        this.monitor = monitor;
+    }
+
+    @Override
+    public synchronized void available()
```

Comment:
It looks like we could have multiple reheats happening at the same time if we have a rapid succession of available/unavailalbe events.

1. `available()` called.
2. `startWarmup()` started, long-running.
3. `unavailable()`, cancel job does nothing with the actual running job. It only prevents starting if it wasn't already started and notifies listeners.
4. `available()` again.
5. `startWarmup()` starts again simultaniously with the old one.

Reply:
Multiple warmups is not a big deal, I think, but I suppose it could lead to multiple profile jobs running at the same time, which would be inconvenient.

Reply:
So this is actually not a problem because the `PageCacheWarmer` methods are `synchronized`. And the `WarmupAvailabilityListener` is likewise `synchronized` to guarantee that we only ever cancel the correct job.